drivers/sms_ser.c

```diff
@@ -0,0 +1,701 @@
+        }
+    }
+
+    results->voltageBattery = atoi(tbattery);
+    results->frequency = atoi(frequency);
+}
+
+void sms_parse_information(uint8_t *rawvalues, SmsData *results) {
+    /* Count from 1 to ignore first char and remove 2 from BUFFER_SIZE
+     *  to compensate the start and ignore '\r' from end. */
+    memset(results->model, 0, sizeof(results->model));
+    memset(results->version, 0, sizeof(results->version));
+
+    for (int i = 1; i < BUFFER_SIZE - 2; i++) {
+        if (i <= 12) {
+            snprintfcat(results->model, 24, "%c", rawvalues[i]);
+        } else {
+            snprintfcat(results->version, 6, "%c", rawvalues[i]);
+        }
+    }
+}
+
+void sms_parse_results(uint8_t *rawvalues, SmsData *results) {
+    char buf[BUFFER_SIZE];
+    uint8_t byte, mask;
+    long v;
+    double h;
+
+    memset(buf, 0, BUFFER_SIZE);
+    sprintf(buf, "0x%02x%02x", (unsigned char)rawvalues[1], (unsigned char)rawvalues[2]);
+    v = strtol(buf, NULL, 16); /* 16 == hex */
+    h = v / 10;
+    results->lastinputVac = h;
+
+    memset(buf, 0, BUFFER_SIZE);
+    sprintf(buf, "0x%02x%02x", (unsigned char)rawvalues[3], (unsigned char)rawvalues[4]);
+    v = strtol(buf, NULL, 16); /* 16 == hex */
+    h = v / 10;
+    results->inputVac = h;
+
+    memset(buf, 0, BUFFER_SIZE);
+    sprintf(buf, "0x%02x%02x", (unsigned char)rawvalues[5], (unsigned char)rawvalues[6]);
+    v = strtol(buf, NULL, 16); /* 16 == hex */
+    h = v / 10;
+    results->outputVac = h;
+
+    memset(buf, 0, BUFFER_SIZE);
+    sprintf(buf, "0x%02x%02x", (unsigned char)rawvalues[7], (unsigned char)rawvalues[8]);
+    v = strtol(buf, NULL, 16); /* 16 == hex */
+    h = v / 10;
+    results->outputpower = h;
+
+    memset(buf, 0, BUFFER_SIZE);
+    sprintf(buf, "0x%02x%02x", (unsigned char)rawvalues[9], (unsigned char)rawvalues[10]);
+    v = strtol(buf, NULL, 16); /* 16 == hex */
+    h = v / 10;
+    results->outputHz = h;
+
+    memset(buf, 0, BUFFER_SIZE);
+    sprintf(buf, "0x%02x%02x", (unsigned char)rawvalues[11], (unsigned char)rawvalues[12]);
+    v = strtol(buf, NULL, 16); /* 16 == hex */
+    h = v / 10;
+    results->batterylevel = h;
+
+    memset(buf, 0, BUFFER_SIZE);
+    sprintf(buf, "0x%02x%02x", (unsigned char)rawvalues[13], (unsigned char)rawvalues[14]);
+    v = strtol(buf, NULL, 16); /* 16 == hex */
+    h = v / 10;
+    results->temperatureC = h;
+
+    byte = rawvalues[15];
+    mask = 1;
+
+    results->beepon = ((byte & (mask << 0)) != 0) ? true : false;
+    results->shutdown = ((byte & (mask << 1)) != 0) ? true : false;
+    results->test = ((byte & (mask << 2)) != 0) ? true : false;
+    results->upsok = ((byte & (mask << 3)) != 0) ? true : false;
+    results->boost = ((byte & (mask << 4)) != 0) ? true : false;
+    results->bypass = ((byte & (mask << 5)) != 0) ? true : false;
+    results->lowbattery = ((byte & (mask << 6)) != 0) ? true : false;
+    results->onbattery = ((byte & (mask << 7)) != 0) ? true : false;
```

Comment:
Wouldn't it be better to define macros for bit mask meanings in the header, and just use `if (rawvalue & SMS_UPSOK) ...` etc.? At least is computationally quicker, uses less memory, and has lots of precedents in NUT code :)

At least, something to consider if/when refactoring this codebase further e.g. to move under `nutdrv_qx` umbrella.